linked_list/linked_list.py

```diff
@@ -1,128 +1,236 @@
+            current = current.next
 
     # method that inserts a given value as a new last node in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(1)
+    # Space Complexity: O(1)
     def add_last(self, value):
-        pass
+        new_node = Node(value)
+
+        if self.head == None:
+            self.head = new_node
+            return new_node
+        else:
+            last = self.head
+            while last.next:
+                last = last.next
+            last.next = new_node
 
     # method to return the max value in the linked list
     # returns the data value and not the node
     def find_max(self):
-        pass
+        current = self.head
+        if current == None:
+            return None
+        max_value = current.value
+        while current != None:
+            if max_value < current.value:
+                max_value = current.value
+            current = current.next
+        return max_value
 
     # method to delete the first node found with specified value
-    # Time Complexity: ?
-    # Space Complexity: ?
-    def delete(self, value):
-        pass
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
+    def delete(self, val):
+        current = self.head
+        previous = None
+
+        while current != None:
+            if current.value == val:
+                if current.next == None:
+                    previous.next = None
+                    return previous
+                current.value = current.next.value
+                current.next = current.next.next
+                continue
+                # return current
+            previous = current
+            current = current.next
+        return current
 
     # method to print all the values in the linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(n)
     def visit(self):
         helper_list = []
         current = self.head
 
         while current:
             helper_list.append(str(current.value))
             current = current.next
-        
+
         print(", ".join(helper_list))
 
     # method to reverse the singly linked list
     # note: the nodes should be moved and not just the values in the nodes
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def reverse(self):
-        pass
-  
+        previous = None
+        current = self.head
+        future = current.next
+
+        while current != None:
+            future = current.next
+            current.next = previous
+            previous = current
+            current = future
+        self.head = previous
+
     ## Advanced/ Exercises
     # returns the value at the middle element in the singly linked list
-    # Time Complexity: ?
-    # Space Complexity: ?
+    # Time Complexity: O(n)
+    # Space Complexity: O(1)
     def find_middle_value(self):

```

Comment:
👍  This works, but you could either:

* use the `length` method
* or have a fast and slow node reference where the fast move 2 nodes with each iteration and the slow moves one.  Then when the fast reaches the end, the slow is referring to the middle node.